kedro/pipeline/node.py

```diff
@@ -397,38 +399,55 @@ def _run_with_dict(self, inputs: Dict[str, Any], node_inputs: Dict[str, str]):
 
     def _outputs_to_dictionary(self, outputs):
         def _from_dict():
-            if set(self._outputs.keys()) != set(outputs.keys()):
+            (result,), iterator = (
+                ((outputs,), None)
+                if not inspect.isgenerator(outputs)
+                else spy(outputs, 1)
+            )
+            keys = list(self._outputs.keys())
+            if not isinstance(result, dict):
+                raise ValueError(
+                    f"Failed to save outputs of node {self}.\n"
+                    f"The node output is a dictionary, whereas the "
+                    f"function output is not."
+                )
+            if set(keys) != set(result.keys()):
                 raise ValueError(
                     f"Failed to save outputs of node {str(self)}.\n"
-                    f"The node's output keys {set(outputs.keys())} do not match with "
-                    f"the returned output's keys {set(self._outputs.keys())}."
+                    f"The node's output keys {set(result.keys())} "
+                    f"do not match with the returned output's keys {set(keys)}."
                 )
-            return {name: outputs[key] for key, name in self._outputs.items()}
+            if iterator:
+                exploded = map(lambda x: tuple(x[k] for k in keys), iterator)
+                result = unzip(exploded)
+            else:
+                result = tuple(result[k] for k in keys)
```

Comment:
I wanted to enforce eager evaluation, but the `dict()` call later will do that for me anyways...

Reply:
@jmholzer Actually we need to evaluate this eagerly due to the fact I am reusing the variable name `result`. Since generators are lazily evaluated, when `dict()` calls indirectly `next()` on the generator, `result` already points to itself, which causes Python to fail with:
```
TypeError: 'generator' object is not subscriptable
```
You've gotten me into a pretty interesting situation of hunting down why it fails when not surrounded by `tuple()` 😆